src/main/java/com/alvarium/utils/SourceCodeUtil.java

```diff
@@ -0,0 +1,133 @@
+
+/*******************************************************************************
+ * Copyright 2021 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+package com.alvarium.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import com.alvarium.hash.HashProvider;
+import com.alvarium.hash.HashProviderFactory;
+import com.alvarium.hash.HashType;
+import com.alvarium.hash.HashTypeException;
+
+public class SourceCodeUtil {
```

Comment:
> I suggest you don't need this class b/c it isn't used in common by anything other than SourceCodeAnnotator. In that sense, it's not a shared "util".

I chose to put it as a shared utility class is to allow the third-party app that consumes the SDK and publishes the `SourceCodeAnnotator` to be able to import this class and use it to generate the checksum. I imagined this would avoid the third-party app being concerned with the implementation details (hash algo, file tree ordering, etc.) of how to generate the checksum.

> Also, when creating a class like this, recommend avoiding overly generic names like "SourceCodeUtil". Consider instead something that indicates what it's used for, like "ChecksumValidator" or "FileChecksumHandler".

Good point. Will implement this change if we decide to keep the separate class.

> Do you see any difficulty with eliminating this class and encapsulating the logic into the SourceCodeAnnotator -- perhaps from a unit test perspective? If not, that's the approach I would recommend.

If you have any ideas on how to eliminate the need for a separate class keeping in mind the concerns I pointed out now, I don't see any issue relating to unit testing or anything else.

Reply:
> I chose to put it as a shared utility class is to allow the third-party app that consumes the SDK and publishes the `SourceCodeAnnotator` to be able to import this class and use it to generate the checksum

I think what you're saying here is that you wanted to expose this for the process that generates the checksum on initial checkout from source control. I don't think that's required since the hashing algorithms are standardized and don't depend on the Java SDK. A DevOps admin might prefer to use some other script to generate the checksum on checkout which is then compared against the product of the SourceCodeAnnotator's methodology directly before the build.

Reply:
Should be resolved in 0991273.